File: _53_Maximum_Subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int ans = Integer.MIN_VALUE;
        int tmp = 0;
        for (int num: nums) {
            tmp += num;
            ans = Math.max(ans, tmp);
            if (tmp < 0) tmp = 0;
        }
        return ans;
    }
}

public class _53_Maximum_Subarray {
    public static void main(String[] args) {
        Solution s = new Solution();
        int[] nums = {-2,1,-3,4,-1,2,1,-5,4};
        System.out.println(s.maxSubArray(nums));
    }
}

File: Python3/53. Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        tmp = nums[0]
        ans = nums[0]
        for v in nums[1:]:
            tmp = tmp + v if tmp + v > v else v
            ans = max(tmp, ans)
        return ans


sol = Solution()
print(sol.maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))
print(sol.maxSubArray([-2]))
